packages/core/src/agent/index.ts

```diff
@@ -1096,97 +889,147 @@ export class Agent<
+        const messageList = new MessageList({ threadId, resourceId, generateMessageId })
+          .addSystem({
+            role: 'system',
+            content: instructions || `${this.instructions}.`,
+          })
+          .add(context || [], 'user');
+
+        if (!memory || (!threadId && !resourceId)) {
+          messageList.add(messages, 'user');
+          return {
+            messageObjects: messageList.get.all.prompt(),
+            convertedTools,
+            messageList,
+          };
+        }
+        if (!threadId || !resourceId) {
+          throw new Error(
+            `A resourceId must be provided when passing a threadId and using Memory. Saw threadId ${threadId} but resourceId is ${resourceId}`,
+          );
+        }
+        const store = memory.constructor.name;
+        this.logger.debug(
+          `[Agent:${this.name}] - Memory persistence enabled: store=${store}, resourceId=${resourceId}`,
+          {
+            runId,
+            resourceId,
+            threadId,
+            memoryStore: store,
+          },
+        );
+
+        const thread =
+          (await memory.getThreadById({ threadId })) ??
+          (await memory.createThread({
+            threadId,
+            resourceId,
+            memoryConfig,
+          }));
+
+        const [memoryMessages, memorySystemMessage] =
+          threadId && memory
+            ? await Promise.all([
+                memory
+                  .rememberMessages({
+                    threadId,
+                    resourceId,
+                    config: memoryConfig,
+                    vectorMessageSearch: messageList.getLatestUserContent() || '',
+                  })
+                  .then(r => r.messages),
+                memory.getSystemMessage({ threadId, memoryConfig }),
+              ])
+            : [[], null];
+
+        this.logger.debug('Fetched messages from memory', {
+          threadId,
+          runId,
+          fetchedCount: memoryMessages.length,
+        });
+
+        if (memorySystemMessage) {
+          messageList.addSystem(memorySystemMessage, 'memory');
+        }
+
+        messageList
+          .add(memoryMessages, 'memory')
+          // add new user messages to the list AFTER remembered messages to make ordering more reliable
+          .add(messages, 'user');
+
+        const systemMessage =
+          messageList
+            .getSystemMessages()
+            ?.map(m => m.content)
+            ?.join(`\n`) ?? undefined;
+
+        const processedMemoryMessages = memory.processMessages({
+          // these will be processed
+          messages: messageList.get.remembered.v1() as CoreMessage[],
+          // these are here for inspecting but shouldn't be returned by the processor
+          // - ex TokenLimiter needs to measure all tokens even though it's only processing remembered messages
+          newMessages: messageList.get.input.v1() as CoreMessage[],
```

Comment:
These are cast to CoreMessage because that's how it was before.. we can improve in a breaking change later